### PR TITLE
Remove PrewhereDAGInfo.

### PR DESCRIPTION
**Pull request by KochetovNicolai**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Changelog category (leave one):
- Not for changelog (changelog entry is not required)

### CODE REVIEW

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 1]
```diff
@@ -978,18 +978,6 @@ void InterpreterSelectQuery::executeImpl(QueryPlan & query_plan, const BlockInpu
 
             prewhere_step->setStepDescription("PREWHERE");
             query_plan.addStep(std::move(prewhere_step));
-
-            // To remove additional columns in dry run
-            // For example, sample column which can be removed in this stage
-            // TODO There seems to be no place initializing remove_columns_actions
```
Comment (amosbird) [inline src/Interpreters/InterpreterSelectQuery.cpp:984]: Nice!

**Comment by KochetovNicolai** [comment]
`01913_names_of_tuple_literal` is considered to be flacky

**Comment by KochetovNicolai** [comment]
Yandex third-party checks should be fixed now.

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 2]
```diff
@@ -1898,28 +1882,7 @@ void InterpreterSelectQuery::executeFetchColumns(QueryProcessingStage::Enum proc
         auto & prewhere_info = analysis_result.prewhere_info;
 
         if (prewhere_info)
```
Comment (alesapin) [inline src/Interpreters/InterpreterSelectQuery.cpp:1884]: Do we need this if now?

File: src/Interpreters/InterpreterSelectQuery.cpp [hunk 3]
```diff
@@ -1479,33 +1467,29 @@ void InterpreterSelectQuery::addEmptySourceToQueryPlan(
 
         if (prewhere_info.alias_actions)
         {
-            pipe.addSimpleTransform(
-                [&](const Block & header) { return std::make_shared<ExpressionTransform>(header, prewhere_info.alias_actions); });
+            pipe.addSimpleTransform([&](const Block & header)
+            {
+                return std::make_shared<ExpressionTransform>(header,
+                    std::make_shared<ExpressionActions>(prewhere_info.alias_actions));
```
Comment (alesapin) [inline src/Interpreters/InterpreterSelectQuery.cpp:1473]: Spent 10 minutes understanding that `alias_actions` is actualy `ActionsDAG`.

File: src/Storages/MergeTree/MergeTreeBaseSelectProcessor.cpp [hunk 4]
```diff
@@ -396,16 +414,17 @@ void MergeTreeBaseSelectProcessor::injectVirtualColumns(
...
 
-void MergeTreeBaseSelectProcessor::executePrewhereActions(Block & block, const PrewhereInfoPtr & prewhere_info)
+Block MergeTreeBaseSelectProcessor::transformHeader(
+    Block block, const PrewhereInfoPtr & prewhere_info, const DataTypePtr & partition_value_type, const Names & virtual_columns)
 {
     if (prewhere_info)
     {
         if (prewhere_info->alias_actions)
-            prewhere_info->alias_actions->execute(block);
+            block = prewhere_info->alias_actions->updateHeader(std::move(block));
```
Comment (alesapin) [inline src/Storages/MergeTree/MergeTreeBaseSelectProcessor.cpp:423]: Why we don't have  `updateHeader(Block &)`?

File: src/Storages/SelectQueryInfo.h [hunk 5]
```diff
@@ -45,34 +42,19 @@ using ClusterPtr = std::shared_ptr<Cluster>;
...
-    bool remove_prewhere_column = false;
-    bool need_filter = false;
-};
-
-/// Same as PrewhereInfo, but with ActionsDAG.
-struct PrewhereDAGInfo
-{
-    ActionsDAGPtr alias_actions;
-    ActionsDAGPtr row_level_filter_actions;
     ActionsDAGPtr prewhere_actions;
```
Comment (alesapin) [inline src/Storages/SelectQueryInfo.h:50]: It's maybe confusing that we have `prewhere_actions` here, but also `PrewhereActions` in RangeReader which has the field `prewhere_actions`...

**Review by alesapin — APPROVED** [review]

File: src/Interpreters/InterpreterSelectQuery.cpp (hunk 2 above)
Comment (KochetovNicolai) [reply src/Interpreters/InterpreterSelectQuery.cpp:1884]: yes ...

**Pull request merged**